pages/components/themeComponents/ThemeProvider.js

```diff
@@ -0,0 +1,61 @@
+import { ThemeProvider as ModuleThemeProvider } from 'styled-components';
+import React from 'react';
+import getTheme from './ThemeColorschemes'
+
+const defaultContextData = {
+  currentTheme: '',
+  toggleTheme: () => { }
+}
+
+const ThemeContext = React.createContext(defaultContextData);
+const currentThemeState = () => React.useContext(ThemeContext);
+
+const ThemeProvider = ({ children }) => {
+  const [themeState, setThemeState] = React.useState({
+    currentTheme: '',
+    hasThemeLoaded: false
+  });
+
+  //useEffect makes it possible to get information from window. Works quite similiar to didComponentsMount() for class react components
+  React.useEffect(() => {
+    const prefersDark = window.matchMedia("(prefers-color-scheme: dark)").matches;
+    const prefersLight = window.matchMedia("(prefers-color-scheme: light)").matches
+
+    const preferedTheme = () => {
+      if (!prefersDark && !prefersLight) {
+        const hourNow = new Date().getHours()
+        return hourNow < 8 || hourNow > 20 ? 'dark' : 'light';
+      }
+      else {
+        return prefersDark ? 'dark' : 'light';
+      }
+    };
+
+    setThemeState({ ...themeState, currentTheme: preferedTheme(), hasThemeLoaded: true });
+  }, [])
+
+  // returns nothing if the theme hasnt been set.
+  // This makes sure we avoid e.g starting in lightmode and switching to dark when we get themeState
+  if (!themeState.hasThemeLoaded) {
+    return null;
```

Comment:
This is something we have to watch out for. The problem with doing it like this is that it breaks server-side rendering. This is the reason I thought it should be set with CSS variables.

In this case the rendering will go something like this:
- The server tries to render the page, but this component will return `null`, and nothing is rendered.
- The client recives no pre-rendered HTML, and starts loading react.
- React does an initial render, but nothing will show, as this still returns `null`.
- The effect runs, and will update `hasThemeLoaded`, so that the page can be loaded.
- The page becomes interactive.

In the case where CSS variables are used:
- The server renders the page independently of which theme that is used.
- The client receives a pre-rendered HTML page for the server.
- The client reads the style sheets, as well as the HTML. The correct theme is already set.
- React starts loading.
- React does an initial render, based on the pre-rendered HTML from the server (the initial renders should be equal).
- The page becomes interactive.

Reply:
can i maybe return an empty div? i looked around a little on different next.js projects on github and saw that some of them returned an empty div if the theme is not loaded

Reply:
That does nothing for the SSR, and the result would be the same.

What we want is to have our entire, full HTML page generated on the server, so that it can be picked up search-engines, links, facebook posts and generally be loaded instantly to the user.

Reply:
The problem with SSR is that i kinda have to use useEffect which kinda does the same as componentsDidMount. Since i cannot get the media queries from the user because there's no "window" during the SSR, hence i wrapped the whole app with styled components "ThemeProvider". This component is also gonna be the root element of everything then basically

Reply:
Created an issue for it #7  and i will eventually get to it )